text/0000-nested-publish.md

```diff
@@ -0,0 +1,303 @@
+* The parent package (and other sub-packages) may have `path =` dependencies upon the sub-package. (Such dependencies must not have a `version =` or `git =`; that is, the `path` must be the _only_ source for the dependency. They do not need to also declare `publish = "nested"`.)
+* You cannot `cargo publish` the sub-package, just as if it had `publish = false`. (This is a safety measure against accidentally publishing the sub-package separately _in addition_ to its nested copies; the presumption here is that nested packages are not designed to present public API themselves.)
+
+Nested packages may contain other dependencies on nested packages, and these too are included in the published package.
+
+When a group of packages is published in this way, this has a number of useful effects for its dependents (which are not things that Cargo explicitly implements, just consequences of the system):
+
+*   The packages are a single unit for all versioning purposes; there is no way for a version mismatch to arise among them since all the code was published together. Version resolution does not apply (in the same way that it does not for any other `path =` dependency).
+
+    This allows package authors to avoid needing to think about SemVer correctness for their nested packages.
+*   The sub-package is effectively “private” (in a sense like the Rust language's visibility system): it cannot be named as a dependency by any other package on `crates.io`, only by its parent package and sibling sub-packages. The parent package may still re-export items from it, or even the entire crate, in the same ways as it could do with a dependency on a normally published package.
+
+## Example: trait and derive macro
+
+Suppose we want to declare a trait-and-derive-macro package. We can do this as follows. The parent package would have this manifest `foo/Cargo.toml`:
+
+```toml
+[package]
+name = "foo"
+version = "0.1.0"
+edition = "2021"
+publish = true
+
+[dependencies]
+foo-macros = { path = "macros", publish = "nested" } # new syntax
+```
+
+The sub-package manifest `foo/macros/Cargo.toml`:
+
+```toml
+[package]
+name = "macros"            # this name need not be claimed on crates.io
+# version = "0.0.0"        # version number is not used and may be omitted
+edition = "2021"
+publish = "nested"         # new syntax
+
+[lib]
+proc-macro = true
+```
+
+Then you can `cargo publish` from within the parent directory `foo/`, and this will create a single `foo` package on `crates.io`, with no `macros` (or `foo-macros`) package visible except when inspecting the source code or in compilation progress messages.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Cargo: Manifest
+
+Two new possible values are added to the manifest.
+
+* The `package.publish` field allows `"nested"` as a value, in addition to existing `false` and `true`. This value affects `cargo publish` and nested dependencies as described below.
```

Comment:
This might be confusing with `package.publish = ["registry"]`.  We might need this to be `package.publish.nested = true` to reduce confusion.

We should probably make it explicit that this value is allowed if `package.version` is missing.

Reply:
Hmmmm. I was not aware of the array being an option. Now that I know it is one, perhaps nestedness _should_ be counted like a registry, since it is a "publishing destination" of sorts? Combining that and your idea, would

```toml
[package]
publish = [{nested = true}, "crates.io"]
```
then make sense as a way to specify that one wanted to permit that duplication?

Reply:
With the sub-table approach, this could also be
```toml
[package]
publish.registries = ["crates.io"]
publish.nested = true
```
I lean towards this approach as I tried to keep the inline data structures simple and I expect this will be easier to reason about than ti fix different publish concepts into the array.

Reply:
I've revised the syntax as you propose. Note that this has the default outcome that

```toml
[package]
publish.registries = true
publish.nested = true
```

is also valid.